lib/logstash/inputs/redis.rb

```diff
@@ -217,6 +237,82 @@ def list_runner(output_queue)
     end
   end
 
+  # private
+  def list_pattern_stop
+    return if @redis.nil? || !@redis.connected?
+
+    @redis.quit rescue nil
+    @redis = nil
+    @threadpool.shutdown
+    @threadpool.wait_for_termination
+  end
+
+  # private
+  def launch_worker(output_queue, key)
+    @logger.debug("Launched worker for #{key}")
+    redis = new_redis_instance
+    begin
+      (0..MAX_ITEMS_PER_WORKER).each do
+        if stop?
+          @logger.debug("Breaking from thread #{key} as it was requested to stop")
+          break
+        end
+        value = redis.lpop(key)
+        break if value.nil?
+        queue_event(value, output_queue)
+        @items_processed.increment
+      end
+    ensure
+      redis.quit rescue nil
+    end
+    @logger.debug("Exit worker for #{key}")
+  end
+
+  # private
+  def threadpool_capacity?
+    @threadpool.remaining_capacity > 0
+  end
+
+  # private
+  def ensure_workers(output_queue)
+    return unless threadpool_capacity?
+    keys = @redis.keys(@key)
+    keys.shuffle
+    keys.each do |key|
+      break unless threadpool_capacity?
+      next if @current_workers.include?(key)
+      @current_workers.add(key)
+      @threadpool.post do
+        begin
+          launch_worker(output_queue, key)
+        ensure
+          @current_workers.delete(key)
+        end
+      end
+    end
+  end
+
+  # private
+  def init_list_pattern_runner
+    init_threadpool
+    @redis ||= connect
+  end
+
+  # private
+  def list_pattern_runner(output_queue)
+    while !stop?
+      begin
+        init_list_pattern_runner if @redis.nil?
+        ensure_workers(output_queue)
+        sleep(0.1)
```

Comment:
Configurable refresh interval might be in order. 0.1s feels quite frequent for listing keys.

Reply:
Yeah maybe. Should probably try to gather some metrics on how this actually performs because whether it's configured or hardcoded, it's just a guess otherwise. Likewise, the sensible number of items processed by a thread should probably be tested somehow.